tutorials/performance/using_multiple_threads.rst

```diff
@@ -188,3 +366,134 @@ ready to be processed:
 
         # Print the counter.
         print("Counter is: ", counter)
+
+ .. code-tab:: cpp C++ .H File
+
+    #ifndef SEMAPHORE_DEMO_H
+    #define SEMAPHORE_DEMO_H
+
+    #include <godot_cpp/classes/mutex.hpp>
+    #include <godot_cpp/classes/node.hpp>
+    #include <godot_cpp/classes/semaphore.hpp>
+    #include <godot_cpp/classes/thread.hpp>
+
+    namespace godot {
+        class SemaphoreDemo : public Node {
+            GDCLASS(SemaphoreDemo, Node);
+
+        private:
+            int counter = 0;
+            Ref<Mutex> mutex;
+            Ref<Semaphore> semaphore;
+            Ref<Thread> thread;
+            bool exit_thread = false;
+
+        protected:
+            static void _bind_methods();
+            void _notification(int p_what);
+
+        public:
+            SemaphoreDemo();
+            ~SemaphoreDemo();
+
+            void thread_function();
+            void increment_counter();
+            int get_counter();
+        };
+    } // namespace godot
+
+    #endif // SEMAPHORE_DEMO_H
+
+ .. code-tab:: cpp C++ .CPP File
+
+    #include "semaphore_demo.h"
+
+    #include <godot_cpp/classes/engine.hpp>
+    #include <godot_cpp/classes/time.hpp>
+    #include <godot_cpp/core/class_db.hpp>
+    #include <godot_cpp/variant/utility_functions.hpp>
+
+    using namespace godot;
+
+    void SemaphoreDemo::_bind_methods() {
+        ClassDB::bind_method(D_METHOD("thread_function"), &SemaphoreDemo::thread_function);
+    }
+
+    void SemaphoreDemo::_notification(int p_what) {
+        // Prevents this from running in the editor, only during game mode.
+        if (Engine::get_singleton()->is_editor_hint()) {
+            return;
+        }
+
+        switch (p_what) {
+            case NOTIFICATION_READY: {
+                UtilityFunctions::print("Counter is starting at: ", counter);
+                mutex.instantiate();
+                semaphore.instantiate();
+                exit_thread = false;
+
+                thread.instantiate();
+                thread->start(callable_mp(this, &SemaphoreDemo::thread_function), Thread::PRIORITY_NORMAL);
+
+                increment_counter(); // Call increment counter to test.
+            } break;
+            case NOTIFICATION_EXIT_TREE, NOTIFICATION_WM_CLOSE_REQUEST: { // Thread must be disposed (or "joined"), for portability.
```

Comment:
```suggestion
            case NOTIFICATION_EXIT_TREE:
            case NOTIFICATION_WM_CLOSE_REQUEST: { // Thread must be disposed (or "joined"), for portability.
```
This isn't valid c++ syntax, and not sure the second part is necessary, if it doesn't crash it will call exit tree before exiting

Reply:
I thought EXIT_TREE should be getting called when X'ing out of the game, but in my testing it doesn't. This was the closest I could find for an answer about the behavior in the Editor: https://github.com/godotengine/godot/issues/45222

We could just replace it with NOTIFICATION_WM_CLOSE_REQUEST but I definitely could be missing something here!

Reply:
That error is different, that's pressing the stop button in the editor, it work when pressing the close button on the window in general, are you sure it doesn't work when pressing the x, not the stop button in the editor?

Reply:
That's a bug to report [here](https://github.com/godotengine/godot-cpp/issues)

But at least the syntax should be fixed 🙂

Reply:
For these examples I could just make all the print commands happen in the functions so that it always works, but I'm a little worried that those close notifications don't seem to be happening consistently. I'm not sure if that's a bug or something I don't understand about running it through the Editor?

Reply:
It could be the prints don't happen correctly, unsure, but let's track that on an issue report instead of here 🙂

Reply:
To confirm, it happens with the corrected code, yes?

Reply:
Tracked it down to: the Output in the Editor won't show the complete output because of this open PR: https://github.com/godotengine/godot/issues/79031

The Console will show correct output.